[PGO][OpenMP] Instrumentation for GPU devices

EthanLuisMcDonough opened the pull request:
This pull request is the first part of an ongoing effort to extends PGO instrumentation to GPU device code. This PR makes the following changes:

- Adds blank registration functions to device RTL
- Gives PGO globals protected visibility when targeting a supported GPU
- Handles any addrspace casts for PGO calls
- Implements PGO global extraction in GPU plugins (currently only dumps info)

These changes can be tested by supplying `-fprofile-instrument=clang` while targeting a GPU.

EthanLuisMcDonough committed:
Add profiling functions to libomptarget

EthanLuisMcDonough committed:
Fix PGO instrumentation for GPU targets

EthanLuisMcDonough committed:
Change global visibility on GPU targets

EthanLuisMcDonough committed:
Make names global public on GPU

EthanLuisMcDonough committed:
Read and print GPU device PGO globals

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

jhuber6 commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -428,13 +428,22 @@ std::string getPGOFuncNameVarName(StringRef FuncName,
   return VarName;
 }
 
+bool isGPUProfTarget(const Module &M) {
+  const auto &triple = M.getTargetTriple();
+  return triple.rfind("nvptx", 0) == 0 || triple.rfind("amdgcn", 0) == 0 ||
+         triple.rfind("r600", 0) == 0;
+}
+
```
```suggestion
bool isGPUProfTarget(const Module &M) {
  const llvm::Triple &Triple = M.getTargetTriple();
  return Triple.isAMDGPU() || Triple.isNVPTX()
}
```
Standard way looks like this. Side note, we really need a way to express this in a more re-usable way especially with SYCL looming. @arsenm should be make some common interface in `CodeGenModule` that just returns if we're currently targeting a "GPU like" device?

EthanLuisMcDonough replied:
I did briefly consider turning the `isGPUProfTarget` function into a method of either the Module or Triple class, but I was worried that would be out of the scope of this PR. Do you think adding an `isGPU` method to the Triple class would make things cleaner?

jhuber6 commented on clang/lib/CodeGen/CodeGenPGO.cpp:
```diff
@@ -959,8 +959,14 @@ void CodeGenPGO::emitCounterIncrement(CGBuilderTy &Builder, const Stmt *S,
-                         Builder.getInt64(FunctionHash),
+  // Make sure that pointer to global is passed in with zero addrspace
+  // This is relevant during GPU profiling
+  auto *I8Ty = llvm::Type::getInt8Ty(CGM.getLLVMContext());
+  auto *I8PtrTy = llvm::PointerType::getUnqual(I8Ty);
+  auto *NormalizedPtr = llvm::ConstantExpr::getPointerBitCastOrAddrSpaceCast(
+      FuncNameVar, I8PtrTy);
+
```
```suggestion
  auto *NormalizedPtr = llvm::ConstantExpr::getPointerBitCastOrAddrSpaceCast(
      FuncNameVar, llvm::PointerType::getUnqual(CGM.getLLVMContext());

```
LLVM uses opaque pointers for everything now.

jhuber6 commented on openmp/libomptarget/DeviceRTL/include/Profiling.h:
```diff
@@ -0,0 +1,21 @@
+
+extern "C" {
+
+void __llvm_profile_register_function(void *ptr);
+void __llvm_profile_register_names_function(void *ptr, long int i);
+}
+
```
```suggestion
void __llvm_profile_register_function(void *Ptr);
void __llvm_profile_register_names_function(void *Ptr, uint64_t I);

}

```

jhuber6 reviewed (commented):
Some quick nits, will look more later.

EthanLuisMcDonough committed:
Fix rebase bug

EthanLuisMcDonough committed:
Refactor portions to be more idiomatic

EthanLuisMcDonough committed:
Reformat DeviceRTL prof functions

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

jdoerfert reviewed (commented):
Can we have tests for this? You can just check for the dump.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/include/GlobalHandler.h:
```diff
@@ -58,6 +60,22 @@ class GlobalTy {
+};
+
+/// PGO profiling data extracted from a GPU device
+struct GPUProfGlobals {
+  std::string names;
+  std::vector<std::vector<int64_t>> counts;
+  std::vector<__llvm_profile_data> data;
+  Triple targetTriple;
+
```
These should probably use LLVM structs. E.g. `StringRef` is the name is a constant string with stable storage and `SmallVector`.

I'd really appreciate some descriptions of how this is supposed to look and how it interacts with the existing profile data.

EthanLuisMcDonough replied:
The basic idea is that the fields in this struct are going to be passed to compiler-rt to be made into a profraw file for this specific target.

I don't know if `llvm::StringRef` would work for the names field. The names data is constant on the GPU device, but dynamic data still has to be allocated on the host in order to read the GPU value. I think it makes the most sense to make the struct own the name data. 

Do you think I should change all the vectors to `SmallVector` or just the innermost `Counts` vector?

jhuber6 replied:
All of them, SmallVector is a std::vector with small size optimizations like `std::string` basically.

EthanLuisMcDonough replied:
I can't seem to use `SmallVector` for `__llvm_profile_data` because many of its defined fields are const qualified and it needs an assignment operator definition. I'm able to use `std::remove_const`, but that feels kind of hack-y. Not sure if the best solution is to use `std::vector`, `std::remove_const`, or remove the const qualifiers in [InstrProfData.inc](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ProfileData/InstrProfData.inc).

jhuber6 replied:
That's confusing, how would using a `std::vector` not have that problem as well? I'll need to look into that.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+  auto ELFObj = getELFObjectFile(Image);
+  if (!ELFObj)
+    return ELFObj.takeError();
+  profdata.targetTriple = ELFObj->makeTriple();
```
Made a patch in https://github.com/llvm/llvm-project/pull/76992 and https://github.com/llvm/llvm-project/pull/76970 to make this actually work.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/include/GlobalHandler.h:
```diff
@@ -58,6 +60,22 @@ class GlobalTy {
   void setPtr(void *P) { Ptr = P; }
 };
 
+typedef void *IntPtrT;
```
What's the utility of this?

EthanLuisMcDonough replied:
`IntPtrT` isn't defined in `profile/InstrProfData.inc`. There are multiple examples of this type being defined. before InstProfData.inc is included. I'm not entirely sure what its for, but I'm assuming its to account for different platforms with different integer sizes.

jhuber6 replied:
Okay. you should use the C++ `using` keyword instead of C's `typedef.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+Expected<GPUProfGlobals>
+GenericGlobalHandlerTy::readProfilingGlobals(GenericDeviceTy &Device,
+                                             DeviceImageTy &Image) {
+  GPUProfGlobals profdata;
```
```suggestion
  GPUProfGlobals ProfData;
```
LLVM style. Also not a fan of the name. Maybe `DeviceProfileData` or something.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+    return ELFObj.takeError();
+  profdata.targetTriple = ELFObj->makeTriple();
+  // Iterate through elf symbols
+  for (auto &sym : ELFObj->symbols()) {
+    if (auto name = sym.getName()) {
+      // Check if given current global is a profiling global based
+      // on name
+      if (name->equals(getInstrProfNamesVarName())) {
+        // Read in profiled function names
+        std::vector<char> chars(sym.getSize() / sizeof(char), ' ');
+        GlobalTy NamesGlobal(name->str(), sym.getSize(), chars.data());
+        if (auto Err = readGlobalFromDevice(Device, Image, NamesGlobal))
+          return Err;
+        std::string names(chars.begin(), chars.end());
+        profdata.names = std::move(names);
+      } else if (name->starts_with(getInstrProfCountersVarPrefix())) {
+        // Read global variable profiling counts
+        std::vector<int64_t> counts(sym.getSize() / sizeof(int64_t), 0);
+        GlobalTy CountGlobal(name->str(), sym.getSize(), counts.data());
+        if (auto Err = readGlobalFromDevice(Device, Image, CountGlobal))
+          return Err;
+        profdata.counts.push_back(std::move(counts));
+      } else if (name->starts_with(getInstrProfDataVarPrefix())) {
+        // Read profiling data for this global variable
+        __llvm_profile_data data{};
+        GlobalTy DataGlobal(name->str(), sym.getSize(), &data);
+        if (auto Err = readGlobalFromDevice(Device, Image, DataGlobal))
+          return Err;
+        profdata.data.push_back(std::move(data));
+      }
+    }
+  }
+  return profdata;
+}
```
LLVM style for everything here.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+  profdata.targetTriple = ELFObj->makeTriple();
+  // Iterate through elf symbols
+  for (auto &sym : ELFObj->symbols()) {
+    if (auto name = sym.getName()) {
```
This is incorrect. If this returns an error it will exit the if, call the deconstructor, and then crash the program because it was not handled.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+      // on name
+      if (name->equals(getInstrProfNamesVarName())) {
+        // Read in profiled function names
+        std::vector<char> chars(sym.getSize() / sizeof(char), ' ');
```
Why are we turning this into a vector of chars? Also isn't `sizeof(char)` pretty much always going to be `1`?

EthanLuisMcDonough replied:
I'm allocating a vector so that I have memory to write the GPU value to.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+      if (name->equals(getInstrProfNamesVarName())) {
+        // Read in profiled function names
+        std::vector<char> chars(sym.getSize() / sizeof(char), ' ');
+        GlobalTy NamesGlobal(name->str(), sym.getSize(), chars.data());
```
Okay, we're reading a string back from the device? What's the purpose of that? Also, just so you know, the ELF will only contain the correct size if it's emitted as an array. E.g.
```
const char a[] = "a"; // strlen("a") + 1 in ELF
const char *b = "b"; // sizeof(char *) in ELF
```

EthanLuisMcDonough replied:
`__llvm_prf_nm` is emitted as an array. For example:

```
@__llvm_prf_nm = protected addrspace(1) constant [63 x i8] c"_=x\DA+I-.\D1K\B6\8A\8F\CF\CF-\88\CFOK\CB\C9OL\C9\CCK\8F7\B1\887M65J41\B5\88\CFM\CC\CC\8B\CF1\B4d,!Z\B1\B1\01X\B1!\00O\F6 \CF", section "__llvm_prf_names", align 1
```

Eventually this data is going to be passed to a function in compiler-rt's profiling library so it can be used to generate a profiling file.

jhuber6 replied:
Okay, this should use `SmallVector` as well, don't bother dividing by the size because the one reported from the ELF is absolute.  Then just make the data inside `uint8_t`.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -163,3 +163,87 @@ Error GenericGlobalHandlerTy::readGlobalFromImage(GenericDeviceTy &Device,
+          return Err;
+        std::string names(chars.begin(), chars.end());
+        profdata.names = std::move(names);
+      } else if (name->starts_with(getInstrProfCountersVarPrefix())) {
```
Are the `getInstrProfCountersVarPrefix` function preexisting? I don't see them defined in this patch set.

EthanLuisMcDonough replied:
Those functions are defined in [InstrProf.h](https://github.com/llvm/llvm-project/blob/597086c60959dd5b3c032552e8b42dd1d053f233/llvm/include/llvm/ProfileData/InstrProf.h#L92-L96).

https://github.com/llvm/llvm-project/blob/597086c60959dd5b3c032552e8b42dd1d053f233/llvm/include/llvm/ProfileData/InstrProf.h#L92-L96

They're used in the creation of PGO global names.

WenleiHe commented:
> ongoing effort to extends PGO instrumentation to GPU device code

Is there a high level description for this effort and its goal? Traditional compiler PGO is mostly for profiling control-flow, but we don't usually have a lot of control flow for GPU kernels.

jdoerfert commented:
> > ongoing effort to extends PGO instrumentation to GPU device code
> 
> Is there a high level description for this effort and its goal? Traditional compiler PGO is mostly for profiling control-flow, but we don't usually have a lot of control flow for GPU kernels.

I am unsure where this assumption comes from but it is not true in general.
HPC most certainly has lots of control flow in kernels. We also have lots of calls, indirect calls, loops of every size, and basically everything else you have in CPU code.

Thus, high level, we want to have PGO for offloaded code for the same reasons we want PGO for non-offloaded codes.

arsenm commented on clang/lib/CodeGen/CodeGenPGO.cpp:
```diff
@@ -959,8 +959,12 @@ void CodeGenPGO::emitCounterIncrement(CGBuilderTy &Builder, const Stmt *S,
+  // Make sure that pointer to global is passed in with zero addrspace
+  // This is relevant during GPU profiling
+  auto *NormalizedPtr = llvm::ConstantExpr::getPointerBitCastOrAddrSpaceCast(
+      FuncNameVar, llvm::PointerType::getUnqual(CGM.getLLVMContext()));
```
Avoid using getUnqual. I'm sure there's a proper address space to query somewhere here

EthanLuisMcDonough replied:
`FuncNameVar` has an addrspace of 1 on GPU devices. The `instrprof_increment` throws a type error if you pass in a pointer without addrspace 0. Is there a reason why one should use `PointerType::get(ctx, 0)` over `PointerType::getUnqual`? Is the latter bad practice/being deprecated?

jdoerfert replied:
@arsenm I think this use is a generic void*, I expect we need/want AS 0 here, always.

arsenm replied:
But generic is a language concept, which doesn't necessarily map to IR AS0. Technically you should go through the language addrspace map

jdoerfert replied:
No, because the runtime, which defines the functions called here, does neither. It expects AS0 calls, which is what we should be generating. If the runtime was "language aware", this might be different.

arsenm commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -448,8 +456,12 @@ GlobalVariable *createPGOFuncNameVar(Module &M,
                          getPGOFuncNameVarName(PGOFuncName, Linkage));
 
+  // If the target is a GPU, make the symbol protected so it can
+  // be read from the host device
+  if (isGPUProfTarget(M))
+    FuncNameVar->setVisibility(GlobalValue::ProtectedVisibility);
   // Hide the symbol so that we correctly get a copy for each executable.
-  if (!GlobalValue::isLocalLinkage(FuncNameVar->getLinkage()))
+  else if (!GlobalValue::isLocalLinkage(FuncNameVar->getLinkage()))
     FuncNameVar->setVisibility(GlobalValue::HiddenVisibility);
```
Can you spot this into some kind of function-to-visibility to use function? This doesn't feel like the right way to control it, so might as well cleanly separate it out

EthanLuisMcDonough committed:
Style changes + catch name error

EthanLuisMcDonough committed:
Add GPU PGO test

EthanLuisMcDonough committed:
Fix PGO test formatting

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Refactor visibility logic

EthanLuisMcDonough committed:
Add LLVM instrumentation support

This PR formerly only supported -fprofile-instrument=clang. This commit adds support for -fprofile-instrument=llvm

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

arsenm commented on llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp:
```diff
@@ -862,14 +862,18 @@ static void instrumentOneFunc(
                                   FuncInfo.FunctionHash);
+  // Make sure that pointer to global is passed in with zero addrspace
+  // This is relevant during GPU profiling
+  auto *NormalizedPtr = llvm::ConstantExpr::getPointerBitCastOrAddrSpaceCast(
+      Name, llvm::PointerType::getUnqual(M->getContext()));
```
Avoid getUnqual, use an explicit 0 addrspace.

Also there should be no reason to emit pointer bitcast, simplify to just create addrspacecast

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Use explicit addrspace instead of unqual

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Remove redundant namespaces

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Clang format

EthanLuisMcDonough committed:
Use getAddrSpaceCast

Replace getPointerBitCastOrAddrSpaceCast with getAddrSpaceCast and allow no-op getAddrSpaceCast calls when types are identical

jhuber6 commented on llvm/lib/IR/Constants.cpp:
```diff
@@ -2067,6 +2067,10 @@ Constant *ConstantExpr::getBitCast(Constant *C, Type *DstTy,
 
 Constant *ConstantExpr::getAddrSpaceCast(Constant *C, Type *DstTy,
                                          bool OnlyIfReduced) {
+  // Skip cast if types are identical
```
Don't think we want to change this function. Was this giving incorrect results without this? Figured a no-op would be valid.

EthanLuisMcDonough replied:
`CastInst::castIsValid` returns false for addrspace casts if two pointers share the same addrspace.

https://github.com/llvm/llvm-project/blob/2de269a641e4ffbb7a44e559c4c0a91bb66df823/llvm/lib/IR/Instructions.cpp#L3802-L3805

arsenm replied:
I guess leave this as it was. The addrspacecast helpers are a mess, we should clear out the vestiges of opaque pointers. The cast build helpers should accept same same address space and elide the cast, like the other cases. Any changes there should be a separate change

EthanLuisMcDonough committed:
Revert "Use getAddrSpaceCast"

This reverts commit 6866862d459e3c3fa65fae8ae639ddc3ff735252.

EthanLuisMcDonough committed:
Fix elf obj file

EthanLuisMcDonough committed:
Add more addrspace casts for GPU targets

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof_ptrcastfix

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Merge branch 'main' into gpuprof

EthanLuisMcDonough committed:
Tighten PGO test requirements

Require compiler-rt to be an enabled runtime

jdoerfert reviewed (approved):
LG, with some nits.

jhuber6 reviewed (commented):
Looks fine in general, I'm not a huge fan of all the `isGPUProfTarget` things we have around now, but I understand it's required to set up the visibility. I wonder if we could factor that out into something more common.

EthanLuisMcDonough committed:
Apply requested formatting changes

EthanLuisMcDonough committed:
Add memop function shim to DeviceRTL

This comes up sometimes when using LLVM IR level instrumentation.

doru1004 commented:
This is failing for me:

```
ld.lld: error: undefined symbol: llvm::InstrProfSymtab::create(llvm::StringRef)
>>> referenced by GlobalHandler.cpp
>>>               GlobalHandler.cpp.o:(llvm::omp::target::plugin::GPUProfGlobals::dump() const) in archive /home/dobercea/upstream/llvm-project/build/lib/libomptarget.rtl.amdgpu.a

ld.lld: error: undefined symbol: llvm::InstrProfSymtab::dumpNames(llvm::raw_ostream&) const
>>> referenced by GlobalHandler.cpp
>>>               GlobalHandler.cpp.o:(llvm::omp::target::plugin::GPUProfGlobals::dump() const) in archive /home/dobercea/upstream/llvm-project/build/lib/libomptarget.rtl.amdgpu.a
clang++: error: linker command failed with exit code 1 (use -v to see invocation)

```

doru1004 commented:
Should this be reverted?

EthanLuisMcDonough commented:
Which platform are you building this on?

doru1004 commented:
I'm building on an x86 + AMD GPU.
What fails is this command:

```
[8/14] Performing build step for 'runtimes'
[1/4] Linking CXX shared library /home/dobercea/upstream/llvm-project/build/lib/libomptarget.so.19.0git
FAILED: /home/dobercea/upstream/llvm-project/build/lib/libomptarget.so.19.0git
: && /home/dobercea/upstream/llvm-project/build/./bin/clang++ --target=x86_64-unknown-linux-gnu -fPIC [....]
```

EthanLuisMcDonough commented:
> I'm building on an x86 + AMD GPU. What fails is this command:
> 
> ```
> [8/14] Performing build step for 'runtimes'
> [1/4] Linking CXX shared library /home/dobercea/upstream/llvm-project/build/lib/libomptarget.so.19.0git
> FAILED: /home/dobercea/upstream/llvm-project/build/lib/libomptarget.so.19.0git
> : && /home/dobercea/upstream/llvm-project/build/./bin/clang++ --target=x86_64-unknown-linux-gnu -fPIC [....]
> ```

Could you share your CMake build command? I'd like to try and replicate this.

doru1004 commented:
Yes of course:

```
cmake \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_INSTALL_PREFIX=~/rocm/trunk_1.0 \
  -DLLVM_ENABLE_PROJECTS="clang;lld;llvm;clang-tools-extra;compiler-rt;flang" \
  -DLLVM_LIT_ARGS="-vv --show-unsupported --show-xfail -j 32" \
  -DLLVM_TARGETS_TO_BUILD="X86;AMDGPU" \
  -DLLVM_ENABLE_ASSERTIONS=ON \
  -DLIBOMPTARGET_ENABLE_DEBUG=ON \
  -DPACKAGE_VENDOR="AMD" \
  -DCLANG_DEFAULT_LINKER=lld \
  -DLLVM_ENABLE_RUNTIMES="openmp;offload" \
  -DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES=libc \
  -DLLVM_RUNTIME_TARGETS="default;amdgcn-amd-amdhsa" \
  -DBUILD_SHARED_LIBS=ON \
  ../llvm -GNinja
```

EthanLuisMcDonough commented:
> Yes of course:
> 
> ```
> cmake \
>   -DCMAKE_BUILD_TYPE=Release \
>   -DCMAKE_INSTALL_PREFIX=~/rocm/trunk_1.0 \
>   -DLLVM_ENABLE_PROJECTS="clang;lld;llvm;clang-tools-extra;compiler-rt;flang" \
>   -DLLVM_LIT_ARGS="-vv --show-unsupported --show-xfail -j 32" \
>   -DLLVM_TARGETS_TO_BUILD="X86;AMDGPU" \
>   -DLLVM_ENABLE_ASSERTIONS=ON \
>   -DLIBOMPTARGET_ENABLE_DEBUG=ON \
>   -DPACKAGE_VENDOR="AMD" \
>   -DCLANG_DEFAULT_LINKER=lld \
>   -DLLVM_ENABLE_RUNTIMES="openmp;offload" \
>   -DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES=libc \
>   -DLLVM_RUNTIME_TARGETS="default;amdgcn-amd-amdhsa" \
>   -DBUILD_SHARED_LIBS=ON \
>   ../llvm -GNinja
> ```

Thanks for this. I'll revert this for now, it seems like the test also fails on the buildbot.

doru1004 commented on offload/test/offloading/pgo1.c:
```diff
@@ -0,0 +1,77 @@
+// LLVM-PGO: ======== Counters =========
+// LLVM-PGO-NEXT: [ 20 ]
+// LLVM-PGO-NEXT: [ 10 ]
+// LLVM-PGO-NEXT: [ 20 10 1 1 ]
```
What do these numbers represent and why have they changed in subsequent commits to 20 10 2 1 ?

doru1004 replied:
@EthanLuisMcDonough
